enhancements/network/user-defined-network-segmentation.md

```diff
@@ -0,0 +1,947 @@
+separate topologies are created whenever a new network is needed. The same methodology will be leveraged for this design.
+Whenever a new network of type layer 3 or layer 2 is requested, a new topology will be created for that network where
+pods may connect to.
+
+The limitation today with secondary networks is that there is only support for east/west traffic. This RFE will address
+adding support for user-defined primary and secondary network north/south support. In order to support north/south
+traffic, pods on different networks need to be able to egress, typically using the host’s IP. Today in shared gateway
+mode we use a Gateway Router (GR) in order to provide this external connectivity, while in local gateway mode, the host
+kernel handles SNAT’ing and routing out egress traffic. Ingress traffic also follows similar, and reverse paths. There
+are some exceptions to these rules:
+
+1. MEG traffic always uses the GR to send and receive traffic.
+2. Egress IP on the primary NIC always uses the GR, even in local gateway mode.
+3. Egress Services always use the host kernel for egress routing.
+
+To provide an ingress/egress point for pods on different networks the most simple solution may appear to be to connect
+them all to a single gateway router. This introduces an issue where now networks are all connected to a single router,
+and there may be routing happening between networks that were supposed to be isolated from one another. Furthermore in
+the future, we will want to extend these networks beyond the cluster, and to do that in OVN would require making a
+single router VRF aware, which adds more complexity into OVN.
+
+The proposal here is to create a GR per network. With this topology, OVN will create a patch port per network to the
+br-ex bridge. OVN-Kubernetes will be responsible for being VRF/network aware and forwarding packets via flows in br-ex
+to the right GR. Each per-network GR will only have load balancers configured on it for its network, and only be able to
+route to pods in its network. The logical topology would look something like this, if we use an example of having a
+cluster default primary network, a layer 3 primary network, and a layer 2 primary network:
+
+![VRF Topology](images/VRFs.svg)
+
+In the above diagram, each network is assigned a unique conntrack zone and conntrack mark. These are required in order
+to be able to handle overlapping networks egressing into the same VRF and SNAT’ing to the host IP. Note, the default
+cluster network does not need to use a unique CT mark or zone, and will continue to work as it does today. This is due
+to the fact that no user-defined network may overlap with the default cluster subnet. More details in the next section.
+
+#### Shared Gateway Mode
+
+##### Pod Egress
+
+On pod egress, the respective GR of that network will handle doing the SNAT to a unique masquerade subnet IP assigned to
```

Comment:
Naif question, Can we differenciate the network using the patch port instead of masquerade address ? this would allow to ditch  masquerade address I think.

Reply:
The masquerade address is still required for avoiding conntrack collisions. See:

https://github.com/openshift/enhancements/pull/1623#discussion_r1623527813